druid/src/menu/mod.rs

```diff
@@ -0,0 +1,762 @@
+        return None;
+
+        // we want to explicitly handle all platforms; log if a platform is missing.
+        tracing::warn!("MenuManager::platform_default is not implemented for this platform.");
+        None
+    }
+
+    /// Called when a menu event is received from the system.
+    pub fn event(&mut self, queue: &mut CommandQueue, id: MenuItemId, data: &mut T, env: &Env) {
+        if let Some(m) = &mut self.menu {
+            let mut ctx = MenuEventCtx { queue };
+            m.activate(&mut ctx, id, data, env);
+        }
+    }
+
+    /// Build an initial menu from the application data.
+    pub fn initialize(&mut self, data: &T, env: &Env) -> PlatformMenu {
+        self.menu = Some((self.build)(data, env));
+        self.old_data = Some(data.clone());
+        self.refresh(data, env)
+    }
+
+    /// Update the menu based on a change to the data.
+    ///
+    /// Returns a new `PlatformMenu` if the menu has changed; returns `None` if it hasn't.
+    pub fn update(&mut self, data: &T, env: &Env) -> Option<PlatformMenu> {
+        if self.old_data.is_none() || self.menu.is_none() {
+            tracing::error!("tried to update uninitialized menus");
+            return None;
+        }
+
+        let old_data = self.old_data.as_ref().unwrap();
+        let menu = self.menu.as_mut().unwrap();
+        let ret = match menu.update(old_data, data, env) {
+            MenuUpdate::NeedsRebuild => {
+                self.menu = Some((self.build)(data, env));
+                Some(self.refresh(data, env))
+            }
+            MenuUpdate::NeedsRefresh => Some(self.refresh(data, env)),
+            MenuUpdate::UpToDate => None,
+        };
+        self.old_data = Some(data.clone());
+        ret
+    }
+
+    pub fn refresh(&mut self, data: &T, env: &Env) -> PlatformMenu {
+        if self.old_data.is_none() || self.menu.is_none() {
+            tracing::error!("tried to refresh uninitialized menus");
+            return PlatformMenu::new();
+        }
+
+        let mut ctx = MenuBuildCtx::new(self.popup);
+        self.menu
+            .as_mut()
+            .unwrap()
+            .refresh_children(&mut ctx, data, env);
+        ctx.current
+    }
+}
+
+/// This context is available to the callback that is called when a menu item is activated.
+///
+/// Currently, it only allows for submission of [`Command`]s.
+///
+/// [`Command`]: crate::Command
+pub struct MenuEventCtx<'a> {
+    queue: &'a mut CommandQueue,
+}
+
+/// This context helps menu items to build the platform menu.
+struct MenuBuildCtx {
+    current: PlatformMenu,
+}
+
+impl MenuBuildCtx {
+    pub(crate) fn new(popup: bool) -> MenuBuildCtx {
+        MenuBuildCtx {
+            current: if popup {
+                PlatformMenu::new_for_popup()
+            } else {
+                PlatformMenu::new()
+            },
+        }
+    }
+
+    pub fn with_submenu(&mut self, text: &str, enabled: bool, f: impl FnOnce(&mut MenuBuildCtx)) {
```

Comment:
It looks like all of these methods are actually internal, in which case we can remove the `pub`/`pub (crate)`, for clarity?